### PR TITLE
feat(n2n): add submit endpoint

### PR DESCRIPTION
**Pull request by Phil91**
## Description

add POST: api/administration/registration/network/partnerRegistration/submit endpoint

## Why

To enable the company admin of the service provider to submit the application

## Issue

N/A - Jira Issue: CPLP-2638

## Depending CD PR

[#88](https://github.com/eclipse-tractusx/portal-cd/pull/88)

## Checklist

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by Phil91** [comment]
> #### ⚠️ GitGuardian has uncovered 5 secrets following the scan of your pull request.
> Please consider investigating the findings and remediating the incidents. Failure to do so may lead to compromising the associated services or software components.
> 
> 🔎 Detected hardcoded secrets in your pull request
> 🛠 Guidelines to remediate hardcoded secrets
> 🦉 [GitGuardian](https://dashboard.gitguardian.com/auth/login/?utm_medium=checkruns&utm_source=github&utm_campaign=cr1) detects secrets in your source code to help developers and security teams secure the modern development process. You are seeing this because you or someone else with access to this repository has authorized GitGuardian to scan your pull request.Our GitHub checks need improvements? [Share your feedbacks](https://050127dayho.typeform.com/to/KmeAPTMk)!

The issue has already been comented. The secret is only used in the unitTests and needed in there because of the encryption which would throw an error if the format isn't matching

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/Administration.Service/Administration.Service.csproj [hunk 1]
```diff
@@ -79,4 +80,8 @@
     </SonarQubeSetting>
   </ItemGroup>
 
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/Administration.Service.csproj]: fixed

File: src/administration/Administration.Service/BusinessLogic/NetworkBusinessLogic.cs [hunk 2]
```diff
@@ -306,4 +306,84 @@ private static void ValidateUsers(UserDetailData user)
...
+        {
+            throw new ConflictException($"Application {companyApplication.Id} is not in state CREATED");
+        }
+
+        if (submitData.Agreements.Any(x => x.ConsentStatusId != ConsentStatusId.ACTIVE))
+        {
+            throw new ConflictException("All agreements must be agreed to");
+        }
+
+        var missingRoles = data.CompanyRoleIds
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/BusinessLogic/NetworkBusinessLogic.cs]: fixed

File: src/administration/Administration.Service/BusinessLogic/NetworkBusinessLogic.cs [hunk 3]
```diff
@@ -306,4 +306,84 @@ private static void ValidateUsers(UserDetailData user)
...
+            .Where(companyRole =>
+                submitData.CompanyRoles.Any(role => role != companyRole.CompanyRoleId));
+        if (missingRoles.Any())
+        {
+            throw new ControllerArgumentException($"CompanyRoles {string.Join(",", missingRoles)} are missing");
+        }
+
+        var allAgreementsActive = data.CompanyRoleIds.SelectMany(x => x.AgreementIds)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/BusinessLogic/NetworkBusinessLogic.cs]: fixed

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs [hunk 4]
```diff
@@ -65,4 +65,43 @@ public NetworkRepository(PortalDbContext context)
                                 step.ProcessStepStatusId == ProcessStepStatusId.TODO))
                 ))
             .SingleOrDefaultAsync();
+
+    public Task<(bool Exists, IEnumerable<(Guid Id, CompanyApplicationStatusId StatusId, string? CallbackUrl)> CompanyApplications, bool IsUserInRole, IEnumerable<(CompanyRoleId CompanyRoleId, IEnumerable<Guid> AgreementIds)> CompanyRoleIds, Guid? ProcessId)> GetSubmitData(Guid companyId, Guid userId, IEnumerable<Guid> roleIds) =>
+        _context.Companies
+            .Where(x => x.Id == companyId)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs:72]: fixed

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs [hunk 5]
```diff
@@ -65,4 +65,43 @@ public NetworkRepository(PortalDbContext context)
...
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.CallbackUrl,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.AuthUrl,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.ClientId,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.ClientSecret),
+                x.ExternalId,
+                x.OnboardingServiceProvider!.BusinessPartnerNumber,
+                x.ApplicationId,
+                processStepTypeId == ProcessStepTypeId.TRIGGER_CALLBACK_OSP_DECLINED ?
+                    x.Process!.ProcessSteps.Where(p => p.ProcessStepTypeId == ProcessStepTypeId.VERIFY_REGISTRATION && p.ProcessStepStatusId == ProcessStepStatusId.FAILED).Select(step => step.Message) :
+                    new List<string?>()))
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs:106]: I had this first, but it fails while executing it.

File: src/administration/Administration.Service/BusinessLogic/RegistrationStatusBusinessLogic.cs [hunk 6]
```diff
@@ -54,20 +60,35 @@ public async Task SetCallbackAddress(OnboardingServiceProviderCallbackData data)
...
         }
 
-        if (ospDetailsExist)
+        var toEncryptedArray = Encoding.UTF8.GetBytes(requestData.ClientSecret);
+        using var aes = Aes.Create();
+        aes.Key = Encoding.UTF8.GetBytes(_settings.EncryptionKey);
+        aes.Mode = CipherMode.ECB;
+        aes.Padding = PaddingMode.PKCS7;
+        using var encryptor = aes.CreateEncryptor();
+        var secretResult = encryptor.TransformFinalBlock(toEncryptedArray, 0, toEncryptedArray.Length);
```
Comment (ntruchsess) [inline src/administration/Administration.Service/BusinessLogic/RegistrationStatusBusinessLogic.cs:69]: I think that call only works reliable when InputBlockSize is taken into account (this is not explicitly documented but one should validate that TransformFinalBlock does support the transformation of a block that is larger than InputBlockSize). From my understanding the incomming bytes should be split into chunks of InputBlockSize and transformed separately while TransformFinalBlock then would transform anything that is left over (being smaller than InputBlockSize).

File: src/portalbackend/PortalBackend.PortalEntities/Entities/OnboardingServiceProviderDetail.cs [hunk 7]
```diff
@@ -25,18 +25,30 @@ public class OnboardingServiceProviderDetail
...
 
     public Guid CompanyId { get; set; }
 
     public string CallbackUrl { get; set; }
 
+    public string AuthUrl { get; set; }
+
+    public string ClientId { get; set; }
+
+    public string ClientSecret { get; set; }
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/Entities/OnboardingServiceProviderDetail.cs:51]: this should be byte[] rendering the conversion to base64 obsolete
(this is supported by postgres, see Document-entity)

File: src/administration/Administration.Service/Models/CompanyRoleConsentDetails.cs [hunk 8]
```diff
@@ -24,7 +24,7 @@
 namespace Org.Eclipse.TractusX.Portal.Backend.Administration.Service.Models;
 
 public record CompanyRoleConsentDetails(
-    [property: JsonPropertyName("companyRoles")] CompanyRoleId CompanyRole,
+    [property: JsonPropertyName("companyRole")] CompanyRoleId CompanyRole,
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/Models/CompanyRoleConsentDetails.cs]: has been reverted

File: src/externalsystems/OnboardingServiceProvider.Library/OnboardingServiceProviderBusinessLogic.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                throw new ArgumentException($"{processStepTypeId} is not supported");
+        }
+
+        var toEncryptArray = Convert.FromBase64String(data.OspDetails.ClientSecret);
+        using var aes = Aes.Create();
+        aes.Key = Encoding.UTF8.GetBytes(_settings.EncryptionKey);
+        aes.Mode = CipherMode.ECB;
+        aes.Padding = PaddingMode.PKCS7;
+        var decryptor = aes.CreateDecryptor();
+        var resultArray = decryptor.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length);
```
Comment (ntruchsess) [inline src/externalsystems/OnboardingServiceProvider.Library/OnboardingServiceProviderBusinessLogic.cs:96]: same as with the encryption part - I expect this should be done blockwise

File: src/portalbackend/PortalBackend.DBAccess/Repositories/CompanyRepository.cs [hunk 10]
```diff
@@ -323,29 +323,36 @@ public void AttachAndModifyProviderCompanyDetails(Guid providerCompanyDetailId,
...
-            .Select(c => new ValueTuple<bool, bool, string?>(
+            .Select(c => new ValueTuple<bool, OspDetails?>(
                 c.CompanyAssignedRoles.Any(role => role.CompanyRoleId == companyRoleId),
-                c.OnboardingServiceProviderDetail != null,
-                c.OnboardingServiceProviderDetail!.CallbackUrl))
+                c.OnboardingServiceProviderDetail == null ?
+                    null :
+                    new OspDetails(c.OnboardingServiceProviderDetail!.CallbackUrl,
+                        c.OnboardingServiceProviderDetail!.AuthUrl,
+                        c.OnboardingServiceProviderDetail!.ClientId,
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/CompanyRepository.cs:341]: ! is not required to be specified more than once for the same nullable

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs [hunk 11]
```diff
@@ -65,4 +65,43 @@ public NetworkRepository(PortalDbContext context)
...
+                    new OspDetails(
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.CallbackUrl,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.AuthUrl,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.ClientId,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.ClientSecret),
+                x.ExternalId,
+                x.OnboardingServiceProvider!.BusinessPartnerNumber,
+                x.ApplicationId,
+                processStepTypeId == ProcessStepTypeId.TRIGGER_CALLBACK_OSP_DECLINED ?
+                    x.Process!.ProcessSteps.Where(p => p.ProcessStepTypeId == ProcessStepTypeId.VERIFY_REGISTRATION && p.ProcessStepStatusId == ProcessStepStatusId.FAILED).Select(step => step.Message) :
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs]: absolutley right, i did filter the 'null' messages

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs [hunk 12]
```diff
@@ -65,4 +65,44 @@ public NetworkRepository(PortalDbContext context)
...
+
+    public Task<(OspDetails? OspDetails, Guid? ExternalId, string? Bpn, Guid ApplicationId, IEnumerable<string?> Comments)> GetCallbackData(Guid networkRegistrationId, ProcessStepTypeId processStepTypeId) =>
+        _context.NetworkRegistrations
+            .Where(x => x.Id == networkRegistrationId)
+            .Select(x => new ValueTuple<OspDetails?, Guid?, string?, Guid, IEnumerable<string?>>(
+                x.OnboardingServiceProvider!.OnboardingServiceProviderDetail == null ?
+                    null :
+                    new OspDetails(
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.CallbackUrl,
+                        x.OnboardingServiceProvider!.OnboardingServiceProviderDetail.AuthUrl,
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs:98]: redundant '!'

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/Administration.Service/BusinessLogic/RegistrationStatusBusinessLogic.cs (hunk 6 above)
Comment (Phil91) [reply src/administration/Administration.Service/BusinessLogic/RegistrationStatusBusinessLogic.cs]: changed it to the usage of a memory and cryptoStream

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs]: adjusted

File: src/portalbackend/PortalBackend.PortalEntities/Entities/OnboardingServiceProviderDetail.cs (hunk 7 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/OnboardingServiceProviderDetail.cs]: changed

File: src/externalsystems/OnboardingServiceProvider.Library/OnboardingServiceProviderBusinessLogic.cs (hunk 9 above)
Comment (Phil91) [reply src/externalsystems/OnboardingServiceProvider.Library/OnboardingServiceProviderBusinessLogic.cs]: changed it to the usage of a memory and cryptoStream

File: src/portalbackend/PortalBackend.DBAccess/Repositories/CompanyRepository.cs (hunk 10 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/CompanyRepository.cs]: removed

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs (hunk 12 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NetworkRepository.cs]: removed

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**